# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/renderer/shade.rs
# Commit: e38dce4a5b06
# Full Hash: e38dce4a5b06869dd4a4d753f41f7594a8cddfb6
# Author: Nicolas Silva <nsilva@mozilla.com>
# Date: 2021-04-22 21:31:57
# Regressor Bug: 1702228
# File Overlap Count: 2
# Description:
#   Bug 1702228 - Cache linear gradients by default. r=gw
#   
#   This patch breaks linear gradients into two primitives:
#    - LinearGradient is always rendered via the brush shader. It is only used with SWGL.
#    - CachedLinearGradient is always rendered via a cached render task and the brush image shader. Its implementation is very similar to conic and radial gradients, with the addition of a fast-path for axis aligned gradients with two stops.
# ==============================================================================

diff -r b3f8befa510f -r e38dce4a5b06 gfx/wr/webrender/src/renderer/shade.rs
--- a/gfx/wr/webrender/src/renderer/shade.rs	Thu Apr 22 10:34:54 2021 +0000
+++ b/gfx/wr/webrender/src/renderer/shade.rs	Thu Apr 22 10:34:54 2021 +0000
@@ -233,6 +233,7 @@
                 VertexArrayKind::Primitive => &desc::PRIM_INSTANCES,
                 VertexArrayKind::LineDecoration => &desc::LINE,
                 VertexArrayKind::FastLinearGradient => &desc::FAST_LINEAR_GRADIENT,
+                VertexArrayKind::LinearGradient => &desc::LINEAR_GRADIENT,
                 VertexArrayKind::RadialGradient => &desc::RADIAL_GRADIENT,
                 VertexArrayKind::ConicGradient => &desc::CONIC_GRADIENT,
                 VertexArrayKind::Blur => &desc::BLUR,
@@ -561,6 +562,7 @@
     pub cs_scale: Vec<Option<LazilyCompiledShader>>,
     pub cs_line_decoration: LazilyCompiledShader,
     pub cs_fast_linear_gradient: LazilyCompiledShader,
+    pub cs_linear_gradient: LazilyCompiledShader,
     pub cs_radial_gradient: LazilyCompiledShader,
     pub cs_conic_gradient: LazilyCompiledShader,
     pub cs_svg_filter: LazilyCompiledShader,
@@ -1010,6 +1012,15 @@
             &shader_list,
         )?;
 
+        let cs_linear_gradient = LazilyCompiledShader::new(
+            ShaderKind::Cache(VertexArrayKind::LinearGradient),
+            "cs_linear_gradient",
+            &[],
+            device,
+            options.precache_flags,
+            &shader_list,
+        )?;
+
         let cs_radial_gradient = LazilyCompiledShader::new(
             ShaderKind::Cache(VertexArrayKind::RadialGradient),
             "cs_radial_gradient",
@@ -1052,6 +1063,7 @@
             cs_border_segment,
             cs_line_decoration,
             cs_fast_linear_gradient,
+            cs_linear_gradient,
             cs_radial_gradient,
             cs_conic_gradient,
             cs_border_solid,
@@ -1254,6 +1266,7 @@
         }
         self.cs_border_solid.deinit(device);
         self.cs_fast_linear_gradient.deinit(device);
+        self.cs_linear_gradient.deinit(device);
         self.cs_radial_gradient.deinit(device);
         self.cs_conic_gradient.deinit(device);
         self.cs_line_decoration.deinit(device);